pkg/agent/flowexporter/exporter/l7Listener.go

```diff
@@ -0,0 +1,189 @@
+		}
+		listener, err := net.Listen("unix", socketPath)
+		if err != nil {
+			return nil, fmt.Errorf("failed to listen on socket: %w", err)
+		}
+		return listener, nil
+	}()
+	if err != nil {
+		klog.Fatalf("Failed to initialize L7Listener: %v", err)
+	}
+	defer listener.Close()
+
+	klog.InfoS("L7 Listener Server started. Listening for connections...")
+	for {
+		select {
+		case <-stopCh:
+			return
+		default:
+			conn, err := listener.Accept()
+			if err != nil {
+				klog.ErrorS(err, "Error accepting Suricata connection")
+				continue
+			}
+			go l.handleClientConnection(conn)
+		}
+	}
+}
+
+func (l *l7Listener) handleClientConnection(conn net.Conn) {
+	defer conn.Close()
+	reader := bufio.NewReader(conn)
+	for {
+		buffer, err := reader.ReadBytes('\n')
+		if err != nil && err != io.EOF {
+			klog.ErrorS(err, "Error reading data", "buffer", buffer)
+			return
+		}
+		err = l.processLog(buffer)
+		if err != nil {
+			klog.ErrorS(err, "Error while processing L7 data")
+			return
+		}
+	}
+}
+
+func (l *l7Listener) processLog(data []byte) error {
+	l.l7mut.Lock()
+	defer l.l7mut.Unlock()
+	var eventType string
+	// Check if the event type is "http"
+	if eventType = extractEventType(data); eventType != "http" {
+		return nil
+	}
+	var event JsonToEvent
+	err := json.Unmarshal(data, &event)
+	if err != nil {
+		return fmt.Errorf("Error parsing JSON data %v", data)
+	}
+	if err = l.addOrUpdateL7EventMap(&event, eventType); err != nil {
+		return fmt.Errorf("Error while adding or Updating L7 event map %v", err)
+	}
+	return nil
+}
+
+func extractEventType(data []byte) string {
+	// Define a temporary struct for extracting the event type
+	type Temp struct {
+		EventType string `json:"event_type"`
+	}
+
+	// Parse the JSON string and extract the event type
+	var temp Temp
+	err := json.Unmarshal(data, &temp)
+	if err != nil && !strings.Contains(err.Error(), "unexpected end of JSON input") {
+		klog.ErrorS(err, "Error parsing JSON for event type:", "data", data)
+		return ""
+	}
+
+	return temp.EventType
+}
+
+func (l *l7Listener) addOrUpdateL7EventMap(event *JsonToEvent, eventType string) error {
+	protocol, err := flowexporter.LookupProtocolMap(event.Proto)
+	if err != nil {
+		return fmt.Errorf("Invalid Protocol type, err: %v", err)
+	}
+	conn := flowexporter.Connection{
+		FlowKey: flowexporter.Tuple{
+			SourceAddress:      event.SrcIP,
+			DestinationAddress: event.DestIP,
+			Protocol:           protocol,
+			SourcePort:         uint16(event.SrcPort),
+			DestinationPort:    uint16(event.DestPort),
+		},
+	}
+	connKey := flowexporter.NewConnectionKey(&conn)
+
+	switch eventType {
+	case "http":
+		if _, ok := l.l7EventsHttp[connKey]; !ok {
+			l.l7EventsHttp[connKey] = make(map[int32]Http)
+		}
+		l.l7EventsHttp[connKey][event.TxID] = event.HTTP
+	}
+	return nil
+}
+
+func (l *l7Listener) getL7EventDataHttpVals(connkey flowexporter.ConnectionKey, sourcePodNN, destinationPodNN string, exp *FlowExporter) string {
```

Comment:
I think there could be a serious memory leak issue if you rely on `getL7EventDataHttpVals` to clean up `l7EventsHttp`. Is it really guaranteed that the function will be called after suricata has sent the event, and suricata will not send another event for this flow after the function is called?

Reply:
Added a PQ for the purpose, if in 15 seconds, the flow didn't ask for the connection key, we will remove the event from l7EventsHttp map

Reply:
I'm thinking why it even attempts to map a L4 connection to a L7 context. I think I commented somewhere before that L7 sessions can reuse a L4 connection. For example, multiple http 1.1 request/responses can reuse a single TCP connection: https://en.wikipedia.org/wiki/HTTP_persistent_connection#HTTP_1.1. It's also what happens today when antrea itself connects to kube-apiserver and antrea-controller and it's perhaps already a widely used pattern of http application. If the L7 visibility module tries to map L4 to L7, then it will get only one L7 event and miss all other L7 events using the same connection.
If the above is agreed, I think instead of relying on L4 connection got from conntrack or OVS, the L7 "connections" got from l7 listener should be sent as an individual flow entry. I think the L7 connection already has the L4 information, so mapping it to K8s context should be viable. And we don't need to worry about the memory leak issue at all as we don't make L4 event rely on L7 event or the reverse.
This is the biggest concern I have so far in the PR.
cc @antoninbas for inputs too.

Reply:
for multiple L7 events, we are already taking care of that as we store them in the same connection key and we populate the flows table whenever it is queried, with then the number of L7 events associated to a connection.

the information we receive from Suricata is 

type Http struct {
	Hostname      string `json:"hostname"`
	URL           string `json:"url"`
	UserAgent     string `json:"http_user_agent"`
	ContentType   string `json:"http_content_type"`
	Method        string `json:"http_method"`
	Protocol      string `json:"protocol"`
	Status        int32  `json:"status"`
	ContentLength int32  `json:"length"`
}

type JsonToEvent struct {
	Timestamp   string  `json:"timestamp"`
	FlowID      int64   `json:"flow_id"`
	InInterface string  `json:"in_iface"`
	EventType   string  `json:"event_type"`
	VLAN        []int32 `json:"vlan"`
	SrcIP       net.IP  `json:"src_ip"`
	SrcPort     int32   `json:"src_port"`
	DestIP      net.IP  `json:"dest_ip"`
	DestPort    int32   `json:"dest_port"`
	Proto       string  `json:"proto"`
	TxID        int32   `json:"tx_id"`
	HTTP        Http    `json:"http"`
}

Creating the identical flows table from this could be challenging. is there something I am missing here ?

Reply:
@tnqn My understanding is that `l7EventsHttp map[flowexporter.ConnectionKey]map[int32]Http` contains all unexported L7 events for the L4 connection, so the case you are describing should be supported. Am I missing something?

For your idea of reversing the logic and exporting a record for each L7 event, would we end up sending one record without L7 data + one record for each L7 event? That sounds like a big departure from how we currently approach things, and it seems that it would put extra burden on the aggregator and the collector?

Reply:
If I understand correctly, the problem is a L4 connection will only be exported once but L7 events can be generated continuously using the same L4 connection. For example, when an application using K8s Golang client lib to access K8s API, they will have only a single TCP connection to kube-apiserver and the application could access different APIs multiple time using the TCP connection during its whole lifycycle unless there is a disconnection. Will we end up exporting the first few request/responses only? And I think suricata will keep sending more L7 events using the connection, how to handle them will be a problem. It would be a memory leak by original code. But even with an expiration mechanism, it would just discard subsequent L7 events as opposed to exporting them.
I'm not sure how collector and aggregator works today, wonder if it's possible to have a L7 level of flow records, which could have the same 5-tuple?

Reply:
The query for the L7 data will only occur for an expired connection and if the connection has expired or not, it will depend on how we take care of our L4 connections.
The case you mentioned is continuous TCP connection over which multiple HTTP connections occur. In the current code, I keep on increasing the timeout for a connection key if the corresponding data is received from Suricata.
and for the case when there is a connection expired and we have the corresponding data in the events map, however, the data is of a group of pods where none are annotated with the direction for which the data has been collected, in that case we should delete the data after the expiration time.
if I am not wrong, I can think of 1 corner case when the timeout has happened however the connection is still active, the issue is there is no way to check if a connection is active or not as the connection key is created by the addOrUpdateL7EventMap using Suricata data and not conntrack data.

Reply:
If L7 data will only be exported upon expiry of L4 connection, could it still lead to huge memory footprint and delayed visibility? All http events using a long-live TCP connection will be buffered in antrea-agent and only be released when the TCP connection terminates (the process exits).

Reply:
> If L7 data will only be exported upon expiry of L4 connection

That's not correct. The connection data should be exported every `activeFlowExportTimeout` (which defaults to 5s). That will include the L7 events received from Suricata since the last export. I feel like we do want to make sure that no events are "lost" at the end of the connection (e.g., can we receive some events from Suricata "late", after the last export?). cc @heanlan

Reply:
What Antonin mentioned is correct. Connections are exported periodically (default 5s) while they are active, and exported once after the connections stop. I seems to me there are some miscommunications above on the term "expired connections/connection expiry". In the context of Flow Exporter, specifically in `exporter.go` file, we use use `expiredConns` to store the connections that should be exported when the 5s timer expired. It doesn't mean those connections are stopped.

In the current PR code, I see the call to `getL7EventDataHttpVals` are made by Flow Exporter before each export. The last time we export a connection is when the connection is [idle timeout](https://github.com/antrea-io/antrea/blob/c82297c5f44bf192662ab765f4bc0fbfd87345c0/build/yamls/antrea.yml#L5721C35-L5721C35) (sorry for another term). If we don't see new packets of this connection during conntrack poll for `idleFlowExportTimeout` (default 15s), then we mark the connection as idle and export it for a last time. Then the question to Tushar is whether the "delay" of Suricata will go beyond 15s? If I understand the problem correctly...

Reply:
Suricata reports events in real time, so I don't believe there could be a delay of 15 seconds. Also I think I should increase the timeout for the l7events by 1 seconds, as last query can occur on 15th second and we can wait for 1 second for query to come, else we will delete the event.

Reply:
I think it is possible that we might receive events from Suricata with a delay after the final export when the TCP state for a connection changes to "CLOSE" or "TIME_OUT"? This might occur because the polling process and the export process operate on different timers. As a result, there's a possibility that the time gap between the TCP state changing to "CLOSE" and the moment of exporting the last record could be relatively short. In such a scenario, is it possible to receive events beyond that point?

Reply:
at [idle timeout](https://github.com/antrea-io/antrea/blob/c82297c5f44bf192662ab765f4bc0fbfd87345c0/build/yamls/antrea.yml#L5721C35-L5721C35), if the connection is closed and we haven't received the event from Suricata, then yes there could be a case where we will miss the event to be written to the flow aggregator, but that seems to be very rare, or am I missing something here? 
The delay from Suricata is minimized by using event driven listener which is currently present in the code.

Reply:
@heanlan when the TCP state goes into `CLOSE` or `TIME_OUT`, can you confirm that we still wait for `idleFlowExportTimeout` before sending the last record?

I think it is reasonable to assume that Suricata won't take more than a few seconds to export events, but you never know with an actual system under load. As long as the probability is very small, I think it should be acceptable for now. I do think that we should log an info message if an L7event is expired from the cache without having been consumed by the exporter.

And regarding the cache, of course we need to have an expiry mechanism to avoid memory leaks. It feels like a reasonable timeout would be `max(activeTimeout, idleTimeout) + 10` seconds, or something along these lines.

Reply:
> when the TCP state goes into CLOSE or TIME_OUT, can you confirm that we still wait for idleFlowExportTimeout before sending the last record

Yes, we do. When we poll and found the TCP state goes into CLOSE or TIME_OUT, we update the connection for the last time and start the `idleFlowExportTimeout` timer.

Reply:
I'm sorry that I forgot about one thing. Please ignore my last comment.

When the TCP state goes into CLOSE or TIME_OUT, the connection will be export once due to `activeFlowExportTimeout`. After this export, the connection will be removed from the priority queue. So it won't be exported for another time due to `idleFlowExportTimeout`. So the wait time between "poll and update TCP state to CLOSE or TIME_OUT" and "last export" is between 0-5s.

Reply:
If it can be close to 0s, then this is probably an issue to consider
Whether it's at the FlowExporter or at the FlogAggregator, we are going to have to deal with the possibility of getting L7 events _after_ we register the TCP connection as "dying" / "dead".
Maybe it is straightforward to hold on to a dying connection for `idleFlowExportTimeout`? @heanlan 

BTW in the above discussion: s/TIME_OUT/TIME_WAIT ...

Reply:
From flow exporter side, I think it is doable. After exporting a connection, if it is dying, then we set its `activeFlowExportTimeout` to be equal to `idleFlowExportTimeout`, and keep it in the PQ. So that we can expect another export when `idleFlowExportTimeout` hit. There should be more changes with the flag `conn.ReadyToDelete` in function [Poll](https://github.com/antrea-io/antrea/blob/c82297c5f44bf192662ab765f4bc0fbfd87345c0/pkg/agent/flowexporter/connections/conntrack_connections.go#L133) otherwise dying connection would be directly deleted from the connection store map.

The problem is on flow aggregator side, FA will skip aggregation of a record if its flowEndSeconds is not greater than the previous one. And it matches this case. Records exported for `activeFlowExportTimeout` and `idleFlowExportTimeout` have the same flowEndSeconds. Unless we do not send the `activeFlowExportTimeout` one? When we found connection is dying, we do not export it for active time out, but only wait for idle time out? It would create a significant delay (15s) for us to receive the last record though.

I'm thinking out load here... please correct me if anything goes wrong...

Reply:
I agree with Anlan, keeping those dying connections might be straightforward on the Flow Exporter side (Just don't delete it in the PQ and the connection map when TCP state changes to "CLOSE" or "TIME_WAIT"). 

The only concern is the reuse of 5-tuple flow key. The conntrack table will be overwritten if we have a new connection with the same 5-tuple flow key within 15s. Since we're using this flow key for the connection map and the PQ, the new connection will be ignored in this case. Although it is an existing issue (not likely to happen), I think it might be worse if we keep the dying connection in PQ or connection map for a longer time? Will holding these dying connections increase the memory usage?

Reply:
> The problem is on flow aggregator side, FA will skip aggregation of a record if its flowEndSeconds is not greater than the previous one. And it matches this case. Records exported for activeFlowExportTimeout and idleFlowExportTimeout have the same flowEndSeconds. Unless we do not send the activeFlowExportTimeout one? When we found connection is dying, we do not export it for active time out, but only wait for idle time out? It would create a significant delay (15s) for us to receive the last record though.

This concern does not exist anymore. I checked the FA code, to update the existing record, we only require the ">=" on `flowEndSeconds` value, not strictly ">". So if we enforce Flow Exporter to send another record with the same `flowEndSeconds` value when idle timeout, FA should not be impacted.